Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx

```diff
@@ -222,87 +201,131 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
 
-  // sort digits by chamber ID for each collision and keep track in index vector
+  // sort digits by half chamber ID for each collision and keep track in index vector
   auto sortStart = std::chrono::high_resolution_clock::now();
-  std::vector<unsigned int> digitIndices(inputDigits.size()); // digit indices sorted by chamber ID for each time frame
-  std::iota(digitIndices.begin(), digitIndices.end(), 0);
-  for (auto& trig : inputTriggerRecords) {
-    std::stable_sort(std::begin(digitIndices) + trig.getFirstDigit(), std::begin(digitIndices) + trig.getNumberOfDigits() + trig.getFirstDigit(),
-                     [&inputDigits](unsigned int i, unsigned int j) { return inputDigits[i].getDetector() < inputDigits[j].getDetector(); });
+  std::vector<unsigned int> digitIdxArray(digits.size()); // digit indices sorted by half chamber ID for each time frame
+  std::iota(digitIdxArray.begin(), digitIdxArray.end(), 0);
+  for (auto& trig : triggerRecords) {
+    std::stable_sort(std::begin(digitIdxArray) + trig.getFirstDigit(), std::begin(digitIdxArray) + trig.getNumberOfDigits() + trig.getFirstDigit(),
+                     [&digits](unsigned int i, unsigned int j) { return digits[i].getHCId() < digits[j].getHCId(); });
   }
-  std::chrono::duration<double> sortTime = std::chrono::high_resolution_clock::now() - sortStart;
-
-  // initialize timers
-  auto timeDigitLoopStart = std::chrono::high_resolution_clock::now();
-  std::chrono::duration<double> trapSimAccumulatedTime{};
-  mTrapSimTime = std::chrono::duration<double>::zero();
-
-  for (int iTrig = 0; iTrig < inputTriggerRecords.size(); ++iTrig) {
-    int nTrackletsInTrigRec = 0;
-    int currDetector = -1;
-    for (int iDigit = inputTriggerRecords[iTrig].getFirstDigit(); iDigit < (inputTriggerRecords[iTrig].getFirstDigit() + inputTriggerRecords[iTrig].getNumberOfDigits()); ++iDigit) {
-      const auto& digit = &inputDigits[digitIndices[iDigit]];
-      if (currDetector < 0) {
-        currDetector = digit->getDetector();
+  auto sortTime = std::chrono::high_resolution_clock::now() - sortStart;
+
+  // prepare data structures for accumulating results per collision
+  std::vector<int> nTracklets(triggerRecords.size());
+  std::vector<std::vector<Tracklet64>> trackletsAccum;
+  trackletsAccum.resize(triggerRecords.size());
+  std::vector<std::vector<short>> digitCountsAccum; // holds the number of digits included in each tracklet (therefore has the same number of elements as trackletsAccum)
+  // digitIndicesAccum holds the global indices of the digits which comprise the tracklets
+  // with the help of digitCountsAccum one can loop through this vector and find the corresponding digit indices for each tracklet
+  std::vector<std::vector<int>> digitIndicesAccum;
+  digitCountsAccum.resize(triggerRecords.size());
+  digitIndicesAccum.resize(triggerRecords.size());
+
+  auto timeParallelStart = std::chrono::high_resolution_clock::now();
+
+#ifdef WITH_OPENMP
+  omp_set_num_threads(mNumThreads);
```

Comment:
Please use `pragma omp parallel for schedule(dynamic) num_threads(n)` instead of `omp_set_num_threads(...)`, since the latter would also affect other omp users.

Reply:
Done. @jolopezl I also applied that change in the TRD digitizer

Reply:
@martenole, thank you. I was going to mention it to you when I read these comments.